system/jlib/jptree.cpp

```diff
@@ -766,6 +1110,113 @@ size32_t CPTValue::queryValueSize() const
 
 ///////////////////
 
+
+CPTArray::~CPTArray()
+{
+    if (map.load())
+        map.load();
+}
+
+CQualifierMap *CPTArray::setMap(CQualifierMap *_map)
+{
+    CQualifierMap *expected = nullptr;
+    if (map.compare_exchange_strong(expected, _map))
+        return nullptr;
+    else
+        return map.load();
+}
+
+void CPTArray::addElement(IPropertyTree *tree)
+{
+    append(*tree);
+    CQualifierMap *map = queryMap();
+    if (map)
+    {
+        if (tree->getAttributeCount())
+        {
+            for (auto &lhs: map->getLhsProps())
+            {
+                const char *lhsp = lhs.c_str();
+                const char *v = tree->queryProp(lhsp);
+                if (v)
+                    map->insertEntry(tree, lhsp, v);
+            }
+        }
+    }
+}
+
+void CPTArray::setElement(unsigned idx, IPropertyTree *tree)
+{
+    CQualifierMap *map = queryMap();
+    if (map)
+    {
+        // remove any mappings for existing element.
+        if (isItem(idx))
+        {
+            IPropertyTree *existing = &((IPropertyTree &)item(idx));
+            for (auto &lhs: map->getLhsProps())
```

Comment:
similarly this should be a member of the map

Reply:
agreed, refactored.